modules/assim.batch/R/pda.bayesian.tools.R

```diff
@@ -125,49 +116,37 @@ pda.bayesian.tools <- function(settings, params.id=NULL, param.names=NULL, prior
   bayesianSetup = createBayesianSetup(bt.likelihood, bt.prior)
   
   
-  ## Set starting values
+  ## Set starting values and numPars
   bayesianSetup$prior$best = parm[prior.ind]
+  bayesianSetup$numPars=n.param
   
-  ## NOTE: The listed samplers here require upper and lower boundaries for now in the BayesianTools package
-  ## check if the user provided them, if not try to extract from the prior if it is uniform, or throw an error
-  upper = as.numeric(unlist(settings$assim.batch$bt.settings$upper))
-  lower = as.numeric(unlist(settings$assim.batch$bt.settings$lower))
-   if(sampler %in% c("DREAM","DREAMzs","M","AM","DR","DRAM")){
-    if(length(upper)!=0 & length(lower)!=0){
-      bayesianSetup$prior$upper=upper
-      bayesianSetup$prior$lower=lower
-    } else if(length(upper)==0 & length(upper)==0 && all(prior[prior.ind,1]=="unif")){
-      bayesianSetup$prior$upper=prior[prior.ind,3]
-      bayesianSetup$prior$lower=prior[prior.ind,2]
-    } else{
-      logger.error(paste0(sampler, " sampler requires upper and lower boundaries for all parameters!"))
-    }
-  }
-  
+
+  logger.info(paste0("Extracting upper and lower boundaries from priors."))
+  rng=matrix(c(sapply(prior.fn$qprior[prior.ind] ,eval,list(p=0.00001)), # M/AM/DR/DRAM can't work with -Inf, Inf values
+               sapply(prior.fn$qprior[prior.ind] ,eval,list(p=0.99999))),
+               nrow=n.param)
+      
+  bayesianSetup$prior$lower=rng[,1]
+  bayesianSetup$prior$upper=rng[,2]
+
   ## Apply BayesianTools specific settings
   bt.settings=pda.settings.bt(settings)
   
   ## central function in BayesianTools
-  out <- runMCMC(bayesianSetup = bayesianSetup, sampler = settings$assim.batch$bt.settings$sampler, settings = bt.settings)
+  out <- runMCMC(bayesianSetup = bayesianSetup, sampler = sampler, settings = bt.settings)
+  
+  # save the out object for further inspection
+  save(out, file=file.path(settings$outdir, paste0(sampler, '_out.Rdata')))
   
-  # save(out,file=file.path(settings$outdir, "out.Rda"))
-       
   ## Create params matrix
-  # *** TODO: Generalize to >1 chain
+  # *** TODO: Generalize to >1 chain, DREAM has 3 chains
   
-  # SHOULD WE USE PACKAGE'S PLOT METHODS?
-  n.row = as.numeric(settings$assim.batch$bt.settings$iter)
+  samples=getSample(out, parametersOnly = T) # getSample{BayesianTools}
+  n.row=length(samples)/n.param
+  params <- matrix(rep(parm,n.row), nrow=n.row, ncol=n.param.all, byrow=T)
+  params[, prior.ind] <- samples
   
-  if(sampler=="SMC"){
-    params <- matrix(rep(parm,bt.settings$initialParticles[[2]]), nrow=bt.settings$initialParticles[[2]], ncol=n.param.all, byrow=T)
```

Comment:
I realized that their getSample function handles different output formats (Metropolis has iteration number +1 rows, DREAM has 3 chains, SMC has particles etc) of each algorithm so I don't have to specify different cases. this version takes care of SMC outputs too